File: reference/conformance/typescript/circuit.ts
/** ShareNet route-to-circuit binding — TypeScript conformance leg (R4-002). */

import { createHash } from "node:crypto";
import type { Value } from "./cbor.ts";
import { encode } from "./cbor.ts";
import { nodeIdentityWire } from "./identity.ts";

const int = (n: bigint): Value => ({ t: "int", v: n });
const txt = (s: string): Value => ({ t: "text", v: s });
const bytes = (b: Uint8Array): Value => ({ t: "bytes", v: b });

function sha256(...parts: Uint8Array[]): Uint8Array {
  const h = createHash("sha256");
  for (const p of parts) h.update(p);
  return new Uint8Array(h.digest());
}

/** CircuitSetup wire: {1: scheme, 2: commitment, 3: nonce, 4: initiator, 5: issued, 6: expires}. */
export function buildSetup(opts: {
  commitmentWire: Uint8Array;
  setupNonce: Uint8Array;
  publicKey: Uint8Array;
  createdAtUnix: bigint;
  issuedAtUnix: bigint;
  validitySecs: bigint;
}): Uint8Array {
  return encode({
    t: "map",
    v: [
      [int(1n), int(1n)],
      [int(2n), bytes(opts.commitmentWire)],
      [int(3n), bytes(opts.setupNonce)],
      [int(4n), nodeIdentityWire(opts.publicKey, opts.createdAtUnix, null)],
      [int(5n), int(opts.issuedAtUnix)],
      [int(6n), int(opts.issuedAtUnix + opts.validitySecs)],
    ],
  });
}

/** CircuitSetupAck wire: {1: scheme, 2: circuit_id, 3: setup_digest, 4: accepting, 5: position, 6: accepted_at, 7: expires_at}. */
export function buildAck(opts: {
  circuitId: Uint8Array;
  setupEnvelope: Uint8Array;
  publicKey: Uint8Array;
  createdAtUnix: bigint;
  position: bigint;
  acceptedAtUnix: bigint;
  validitySecs: bigint;
}): Uint8Array {
  return encode({
    t: "map",
    v: [
      [int(1n), int(1n)],
      [int(2n), bytes(opts.circuitId)],
      [int(3n), bytes(sha256(opts.setupEnvelope))],
      [int(4n), nodeIdentityWire(opts.publicKey, opts.createdAtUnix, null)],
      [int(5n), int(opts.position)],
      [int(6n), int(opts.acceptedAtUnix)],
      [int(7n), int(opts.acceptedAtUnix + opts.validitySecs)],
    ],
  });
}

/** CircuitFrame wire: {1: scheme, 2: circuit_id, 3: direction, 4: seq, 5: payload}. */
export function buildFrame(opts: {
  circuitId: Uint8Array;
  direction: bigint;
  seq: bigint;
  payload: Uint8Array;
}): Uint8Array {
  return encode({
    t: "map",
    v: [
      [int(1n), int(1n)],
      [int(2n), bytes(opts.circuitId)],
      [int(3n), int(opts.direction)],
      [int(4n), int(opts.seq)],
      [int(5n), bytes(opts.payload)],
    ],
  });
}

/** CircuitDestroy wire: {1: scheme, 2: circuit_id, 3: sender, 4: reason, 5: destroyed_at}. */
export function buildDestroy(opts: {
  circuitId: Uint8Array;
  publicKey: Uint8Array;
  createdAtUnix: bigint;
  reason: string;
  destroyedAtUnix: bigint;
}): Uint8Array {
  return encode({
    t: "map",
    v: [
      [int(1n), int(1n)],
      [int(2n), bytes(opts.circuitId)],
      [int(3n), nodeIdentityWire(opts.publicKey, opts.createdAtUnix, null)],
      [int(4n), txt(opts.reason)],
      [int(5n), int(opts.destroyedAtUnix)],
    ],
  });
}

/** circuit_id = SHA-256("sharenet-circuit-id-v1" || route_id || setup_nonce). */
export function deriveCircuitId(routeId: Uint8Array, setupNonce: Uint8Array): Uint8Array {
  return sha256(
    new TextEncoder().encode("sharenet-circuit-id-v1"),
    routeId,
    setupNonce,
  );
}

/** The ack binding digest: SHA-256 of the signed setup envelope bytes. */
export function setupDigest(setupEnvelope: Uint8Array): Uint8Array {
  return sha256(setupEnvelope);
}

File: reference/conformance/python/sharenet_conformance/circuit.py
"""ShareNet route-to-circuit binding — Python conformance leg (R4-002)."""

import hashlib

from .cbor import CborMap, encode


def _identity(pub: bytes, created_at_unix: int) -> CborMap:
    # nested map (NOT the encoded identity wire bytes)
    return CborMap([(1, 1), (2, bytes(pub)), (3, created_at_unix)])


def build_setup(commitment_wire: bytes, setup_nonce: bytes, public_key: bytes,
                created_at_unix: int, issued_at_unix: int, validity_secs: int) -> bytes:
    """CircuitSetup wire: {1: scheme, 2: commitment, 3: nonce, 4: initiator, 5: issued, 6: expires}."""
    return encode(
        CborMap(
            [
                (1, 1),
                (2, commitment_wire),
                (3, setup_nonce),
                (4, _identity(public_key, created_at_unix)),
                (5, issued_at_unix),
                (6, issued_at_unix + validity_secs),
            ]
        )
    )


def build_ack(circuit_id: bytes, setup_envelope: bytes, public_key: bytes,
              created_at_unix: int, position: int, accepted_at_unix: int,
              validity_secs: int) -> bytes:
    """CircuitSetupAck wire: {1: scheme, 2: circuit_id, 3: setup_digest, 4: accepting,
    5: position, 6: accepted_at, 7: expires_at}."""
    return encode(
        CborMap(
            [
                (1, 1),
                (2, circuit_id),
                (3, hashlib.sha256(setup_envelope).digest()),
                (4, _identity(public_key, created_at_unix)),
                (5, position),
                (6, accepted_at_unix),
                (7, accepted_at_unix + validity_secs),
            ]
        )
    )


def build_frame(circuit_id: bytes, direction: int, seq: int, payload: bytes) -> bytes:
    """CircuitFrame wire: {1: scheme, 2: circuit_id, 3: direction, 4: seq, 5: payload}."""
    return encode(
        CborMap(
            [
                (1, 1),
                (2, circuit_id),
                (3, direction),
                (4, seq),
                (5, payload),
            ]
        )
    )


def build_destroy(circuit_id: bytes, public_key: bytes, created_at_unix: int,
                  reason: str, destroyed_at_unix: int) -> bytes:
    """CircuitDestroy wire: {1: scheme, 2: circuit_id, 3: sender, 4: reason, 5: destroyed_at}."""
    return encode(
        CborMap(
            [
                (1, 1),
                (2, circuit_id),
                (3, _identity(public_key, created_at_unix)),
                (4, reason),
                (5, destroyed_at_unix),
            ]
        )
    )


def derive_circuit_id(route_id: bytes, setup_nonce: bytes) -> bytes:
    """circuit_id = SHA-256("sharenet-circuit-id-v1" || route_id || setup_nonce)."""
    return hashlib.sha256(
        b"sharenet-circuit-id-v1" + route_id + setup_nonce
    ).digest()


def commitment_wire(proposal_env: bytes, acceptance_envs, root: bytes, route_id: bytes) -> bytes:
    """RouteCommitment wire: {1: scheme, 2: proposal env, 3: acceptance envs,
    4: root, 5: route_id}."""
    return encode(
        CborMap(
            [
                (1, 1),
                (2, proposal_env),
                (3, list(acceptance_envs)),
                (4, root),
                (5, route_id),
            ]
        )
    )
